File: go/internal/tools/hello.go
// Package tools provides built-in MCP tools for LinodeMCP.
package tools

import (
	"context"
	"fmt"

	"github.com/mark3labs/mcp-go/mcp"

	"github.com/chadit/LinodeMCP/go/internal/config"
	linodev1 "github.com/chadit/LinodeMCP/go/internal/genpb/linode/mcp/v1"
	"github.com/chadit/LinodeMCP/go/internal/profiles"
	"github.com/chadit/LinodeMCP/go/internal/toolschemas"
)

// NewHelloTool creates a hello tool for smoke testing.
func NewHelloTool(_ *config.Config) (mcp.Tool, profiles.Capability, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error)) {
	tool := mcp.NewToolWithRawSchema(
		"hello",
		"Responds with a friendly greeting from LinodeMCP",
		toolschemas.Schema("linode.mcp.v1.HelloInput"),
	)

	handler := func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		default:
		}

		name := request.GetString("name", "World")

		return MarshalProtoToolResponse(&linodev1.HelloResponse{
			Message: fmt.Sprintf("Hello, %s! LinodeMCP server is running and ready.", name),
		})
	}

	return tool, profiles.CapMeta, handler
}

File: python/src/linodemcp/tools/hello.py
"""Hello tool - friendly greeting."""

import json
from typing import Any

from mcp.types import TextContent, Tool

from linodemcp.genpb.linode.mcp.v1 import version_pb2
from linodemcp.profiles import Capability
from linodemcp.tools.proto_response import proto_to_canonical_dict
from linodemcp.tools.toolschemas import schema


def create_hello_tool() -> tuple[Tool, Capability]:
    """Create the hello tool."""
    return Tool(
        name="hello",
        description="Responds with a friendly greeting from LinodeMCP",
        inputSchema=schema("linode.mcp.v1.HelloInput"),
    ), Capability.Meta


async def handle_hello(arguments: dict[str, Any]) -> list[TextContent]:
    """Handle hello tool request.

    Args:
        arguments: HelloArgs - name (optional)
    """
    name = arguments.get("name", "World")
    message = version_pb2.HelloResponse(
        message=f"Hello, {name}! LinodeMCP server is running and ready."
    )
    return [
        TextContent(
            type="text", text=json.dumps(proto_to_canonical_dict(message), indent=2)
        )
    ]
